### PR TITLE
feat: add support for RTK-related functionalities

### PR DESCRIPTION
**Pull request by doudou**
Depends on:
- [x] https://github.com/rock-drivers/drivers-gps_base/pull/14

### CODE REVIEW

**Review by rodrigoleonello — APPROVED** [review]

File: src/GPSData.hpp [hunk 1]
```diff
@@ -11,7 +11,6 @@ namespace gps_ublox {
      * GPS data
      */
     struct GPSData {
-
```
Comment (caioaamaral) [inline src/GPSData.hpp:14]: is there a specific rule for line breaks?

File: src/Main.cpp [hunk 2]
```diff
@@ -19,6 +20,18 @@ static const map<string, Driver::DeviceProtocol> PROTOCOL_S_TO_E = {
     { "rtcm3x", Driver::PROTOCOL_RTCM3X }
 };
 
+static const map<GPSData::GNSSFixType, string> FIX_TO_STRING = {
```
Comment (caioaamaral) [inline src/Main.cpp]: FIX_TYPE_TO_STRING?

File: src/Main.cpp [hunk 3]
```diff
@@ -130,6 +144,28 @@ int main(int argc, char** argv)
...
+                if ((flag.first & data.fix_flags) == flag.first) {
+                    fix_type += "/" + flag.second;
+                }
+            }
+
+            std::cout
+                << data.latitude.getDeg() << " " << data.longitude.getDeg() << " "
+                << data.height << " " << data.height_above_mean_sea_level << " "
+                << fix_type << "\n";
+        }
```
Comment (caioaamaral) [inline src/Main.cpp]: Maybe I did miss something, but how does it exit the while loop? Shouldn't we check if `data` is valid?

File: CMakeLists.txt [hunk 4]
```diff
@@ -1,7 +1,10 @@
 # CMakeLists.txt has to be located in the project folder and cmake has to be
 # executed from 'project/build' with 'cmake ../'.
-cmake_minimum_required(VERSION 2.6)
+cmake_minimum_required(VERSION 3.9)
```
Comment (caioaamaral) [inline CMakeLists.txt:3]: why do we need this specific version as the minimum required?

File: src/cfg.hpp [hunk 5]
```diff
@@ -0,0 +1,85 @@
+#ifndef GPS_UBLOX_CONFIGURATION_HPP
+#define GPS_UBLOX_CONFIGURATION_HPP
+
+#include <cstdint>
+
+namespace gps_ublox {
+    /** Identifies the device's ports
+     */
+    enum DevicePort {
```
Comment (caioaamaral) [inline src/cfg.hpp:9]: why do not use `enum class` in this case?

File: src/Driver.cpp [hunk 6]
```diff
@@ -11,6 +12,7 @@ using namespace std;
 using namespace gps_ublox;
 using namespace UBX;
 
+
```
Comment (caioaamaral) [inline src/Driver.cpp]: why the line break?

File: src/cfg.cpp [hunk 7]
```diff
@@ -41,3 +41,42 @@ static int getMsgOutPortOffset(DevicePort port) {
 std::uint32_t cfg::getOutputRateKey(DevicePort port, MessageOutputType type) {
     return getMsgOutPortOffset(port) + type;
 }
+
+struct RTCMConfig {
+    static const int BASE = 0x20910000;
```
Comment (caioaamaral) [inline src/cfg.cpp]: I think `constexpr` would fit here

File: src/cfg.cpp [hunk 8]
```diff
@@ -41,3 +41,42 @@ static int getMsgOutPortOffset(DevicePort port) {
...
+    { 1087, 0x2d1 },
+    { 1094, 0x368 },
+    { 1097, 0x318 },
+    { 1124, 0x36d },
+    { 1127, 0x2d6 },
+    { 1230, 0x303 },
+    { 4072, 0x2fe },
+    { 40720, 0x2fe },
+    { 40721, 0x381 }
+};
```
Comment (caioaamaral) [inline src/cfg.cpp]: Just a piece of random information, for this use case `RTCMConfig` does not need a user-defined constructor https://godbolt.org/z/KoWaGajo1

File: src/Main.cpp [hunk 9]
```diff
@@ -1,22 +1,54 @@
...
+    { "usb", PORT_USB }
+};
+
+static const map<string, DeviceProtocol> PROTOCOL_S_TO_E = {
+    { "ubx", PROTOCOL_UBX },
+    { "nmea", PROTOCOL_NMEA },
+    { "rtcm3x", PROTOCOL_RTCM3X }
+};
+
+static const map<PVT::GNSSFixType, string> FIX_TO_STRING = {
```
Comment (caioaamaral) [inline src/Main.cpp]: FIX_TYPE_TO_STRING?

**Review by caioaamaral — APPROVED** [review]
All good! Just made a few questions (I'm more curious about the decisions than suggesting any change)

File: src/Main.cpp (hunk 3 above)
Comment (caioaamaral) [reply src/Main.cpp]: got it, just ignore this one

File: src/cfg.hpp (hunk 5 above)
Comment (doudou) [reply src/cfg.hpp:9]: Because orogen does not support `enum class`

File: src/Driver.cpp (hunk 6 above)
Comment (doudou) [reply src/Driver.cpp]: Because there isn't a linter to warn me about it

File: CMakeLists.txt (hunk 4 above)
Comment (doudou) [reply CMakeLists.txt:3]: project( ... DESCRIPTION ...)

File: src/cfg.cpp (hunk 7 above)
Comment (doudou) [reply src/cfg.cpp]: Ah ... young people ... :P

It would. I guess it's a matter of habit.

File: src/cfg.cpp (hunk 8 above)
Comment (doudou) [reply src/cfg.cpp]: This is true. I created the constructor during the development but then forgot to remove it.

Comment (doudou) [reply src/cfg.cpp]: Actually, only partially true. Does not work with C++11 if we retain the field initialization (which is what this code is compiled with).

File: src/Main.cpp (hunk 9 above)
Comment (doudou) [reply src/Main.cpp]: Done

**Pull request merged**